### PR TITLE
Restore sessions after Android process death

### PR DESCRIPTION
**Pull request by corigne**
This is the follow-up I mentioned in #1624.

When Android kills melonDS while the device is sleeping, reopening the app now restores a private checkpoint instead of returning to the ROM list or restarting the BIOS. Automatic recovery is limited to the exact previous process and Android low-memory, resource, freezer, or SIGKILL exits. Prompted recovery disables Hardcore.

I reproduced the original path on an AYN Thor running Android 13 by forcing real LMKD pressure while the device was asleep. Android reported `LOW_MEMORY`; the ROM restored without a dialog, Hardcore remained enabled, animation and touch worked, and the next sleep checkpoint completed normally. Firmware recovery was also tested through the animated health screen and into the firmware menu.

The recovery policy and event framing have portable JVM tests run by a dedicated pull-request workflow.

### CODE REVIEW

File: app/src/main/java/me/magnum/melonds/ui/emulator/LidCloseService.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        private const val NOTIFICATION_ID_LID_CLOSE = 300
+    }
+
+    override fun onBind(intent: Intent?): IBinder? = null
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        val notification = NotificationCompat.Builder(this, MelonDSApplication.NOTIFICATION_CHANNEL_ID_BACKGROUND_TASKS)
+            .setContentTitle("DS Sleep")
+            .setSmallIcon(android.R.drawable.ic_lock_silent_mode)
+            .setSilent(true)
```
[the remark this answers is not in the document]
Comment (corigne) [reply app/src/main/java/me/magnum/melonds/ui/emulator/LidCloseService.kt]: Outdated after isolating this PR from the lid changes. LidCloseService is no longer part of this branch.

File: app/src/main/java/me/magnum/melonds/impl/emulator/EmulatorMessageQueue.kt [hunk 2]
```diff
@@ -106,33 +108,105 @@ class EmulatorMessageQueue(private val eventHandler: EventHandler) {
...
+        messagesFileDescriptor?.let { fd ->
+            Looper.myLooper()?.queue?.removeOnFileDescriptorEventListener(fd.fileDescriptor)
+            fd.close()
+        }
 
-        eventBuffer.clear()
-        var bytesRead = currentInputStream.channel.read(eventBuffer)
-        eventBuffer.position(0)
+        inputStream = null
+        messagesFileDescriptor = null
```
[the remark this answers is not in the document]
Comment (corigne) [reply app/src/main/java/me/magnum/melonds/impl/emulator/EmulatorMessageQueue.kt:124]: Fixed in 5416d16b. The stream now owns the ParcelFileDescriptor through AutoCloseInputStream and is closed during shutdown.

File: app/src/main/java/me/magnum/melonds/impl/emulator/recovery/EmulatorRecoveryRepository.kt [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+            }
+            val entry = JSONObject()
+                .put("timestamp", System.currentTimeMillis())
+                .put("elapsedRealtime", SystemClock.elapsedRealtime())
+                .put("processId", Process.myPid())
+                .put("processToken", processToken)
+                .put("event", event)
+                .put("details", details)
+            journalFile.appendText(entry.toString() + "\n")
+            Log.i(TAG, "$event: $details")
```
[the remark this answers is not in the document]
Comment (corigne) [reply app/src/main/java/me/magnum/melonds/impl/emulator/recovery/EmulatorRecoveryRepository.kt]: Fixed in 5416d16b. Logcat now receives only the event name; details remain in the app-private journal.

File: app/src/main/java/me/magnum/melonds/impl/emulator/recovery/EmulatorRecoveryRepository.kt [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+
+    private fun isCheckpointValid(session: RecoverySession): Boolean {
+        if (session.appVersionCode != currentVersionCode()) {
+            return false
+        }
+        val expectedChecksum = session.checkpointSha256 ?: return false
+        val checkpointFile = getCheckpointFile(session) ?: return false
+        return checkpointFile.isFile &&
+            checkpointFile.length() > 0L &&
+            checkpointFile.sha256() == expectedChecksum
```
[the remark this answers is not in the document]
Comment (corigne) [reply app/src/main/java/me/magnum/melonds/impl/emulator/recovery/EmulatorRecoveryRepository.kt:407]: Fixed in 5416d16b. Pending recovery validation and automatic-recovery persistence now run on Dispatchers.IO.

**Comment by corigne** [comment]
Fixed a wake regression in `b5cb55f7`. When checkpoint work had already completed, `join()` returned inline from `onResume()` before lifecycle state reached `RESUMED`, so the guard aborted the resume path. The continuation now yields once before checking lifecycle state.

Verified on the AYN Thor with both the recovery-only branch and the combined lid/RTC build: screen off, checkpoint, screen on, `device_sleep_resumed`, responsive BIOS/game output.

---

## android backup issue (and fix)

Reproduced on an AYN Thor running Android 13 with `me.magnum.melonds.dev` 2.0.1 GH (41). 

Ran a test overnight, Android selected the sleeping app for full backup; backup completed, then PID 13879 died and `LidCloseService` stopped at 01:07:42. Android recorded the kill reason as not an LMKD memory pressure based kill. The recovery record remained active and sleeping with a valid checkpoint and `automaticRecoveryAttempted=false`.

A couple of hours later, another full backup started PID 30866 as `{android/FullBackupAgent}` while Android simultaneously restarted the existing `EmulatorActivity` task in that process. The activity failed before recovery initialization with `Hilt Activity must be attached to an @HiltAndroidApp Application`; Android force-finished it, cancelled backup, and started the underlying `RomDetailsActivity`. The added recovery functiaonlity never had a chance to run. The installed manifest did contain `android:name="me.magnum.melonds.MelonDSApplication"`; this was a backup/task-restoration race. Unfortunately, I'm not familiar with any way to combat a backup race than disabling backups for the package.

Fix: f846867b disables app backup with `android:allowBackup="false"` in the manifest. Recovery checkpoints are runtime continuity state and are incompatible with Auto Backup. It may be possible to exfiltrate that state to some stable storage, but I have not found a clean way to do so. It is worth disuccsing.  

96d00a8d also restores automatically for every known non-user indicated process exit. Explicit user-initiated stops are discarded without a dialog, while rare edge cases/unknown causes retain the explicit fallback prompt. 

549240fa updates the test guide.

Verification:
- `RecoveryPolicyTest` and `EmulatorEventFrameDecoderTest`: 12 tests passed, 0 failed.
- `:app:processGitHubProdDebugMainManifest`: passed; packaged `GitHubProdDebug` manifest contains `android:name="me.magnum.melonds.MelonDSApplication"` and `android:allowBackup="false"`.
- Immediate isolated-package test with a running game reproduced the pre-fix failure: screen off committed a checkpoint, `bmgr backupnow` started full backup, killed the emulator process, and activity restoration in the backup process hit the Hilt failure.
- The patched package returned `Backup is not allowed`; its PID survived and wake recorded `device_sleep_resumed` with `emulationResumed=true`.
- A fresh patched session was checkpointed, deliberately crashed, and relaunched. It restored automatically without a prompt and recorded `recovery_restored` with `automatic=true`.

<details>
<summary>Sanitized log: backup stopped the sleeping process</summary>

```text
09-08 01:07:40.430 I full_backup_package: me.magnum.melonds.dev
09-08 01:07:41.971 I full_backup_success: me.magnum.melonds.dev
09-08 01:07:42.039 I am_proc_died: [0,13879,me.magnum.melonds.dev,200,4]
09-08 01:07:42.045 I am_foreground_service_stop: [0,me.magnum.melonds.dev/me.magnum.melonds.ui.emulator.LidCloseService,...,STOP_SERVICE]
```

</details>

<details>
<summary>Sanitized log: backup process restarted the emulator task</summary>

```text
09-08 03:13:35.991 I full_backup_package: me.magnum.melonds.dev
09-08 03:13:36.000 I am_proc_start: [0,30866,...,me.magnum.melonds.dev,backup,{android/FullBackupAgent}]
09-08 03:13:36.000 I ActivityManager: Start proc 30866:me.magnum.melonds.dev/... for backup {android/FullBackupAgent}
09-08 03:13:36.037 I wm_restart_activity: [0,...,...,me.magnum.melonds.dev/me.magnum.melonds.ui.emulator.EmulatorActivity]
09-08 03:13:36.388 E AndroidRuntime: FATAL EXCEPTION: main
09-08 03:13:36.388 E AndroidRuntime: Process: me.magnum.melonds.dev, PID: 30866
09-08 03:13:36.388 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{me.magnum.melonds.dev/me.magnum.melonds.ui.emulator.EmulatorActivity}: java.lang.IllegalStateException: Hilt Activity must be attached to an @HiltAndroidApp Application
09-08 03:13:36.388 E AndroidRuntime: Caused by: java.lang.IllegalStateException: Hilt Activity must be attached to an @HiltAndroidApp Application
09-08 03:13:36.388 E AndroidRuntime:     at dagger.hilt.android.internal.managers.ActivityComponentManager.createComponent(ActivityComponentManager.java:102)
09-08 03:13:36.388 E AndroidRuntime:     at me.magnum.melonds.ui.emulator.Hilt_EmulatorActivity.generatedComponent(Hilt_EmulatorActivity.java:67)
09-08 03:13:36.388 E AndroidRuntime:     at me.magnum.melonds.ui.emulator.EmulatorActivity.onCreate(EmulatorActivity.kt:314)
09-08 03:13:36.388 E AndroidRuntime:     [truncated]
09-08 03:13:36.392 W ActivityTaskManager: Force finishing activity me.magnum.melonds.dev/me.magnum.melonds.ui.emulator.EmulatorActivity
09-08 03:13:36.488 W BackupManagerService: agentDisconnected: the backup agent for me.magnum.melonds.dev died: cancel current operations
09-08 03:13:36.489 W PFTBT: Full backup cancel of me.magnum.melonds.dev
09-08 03:13:36.498 I full_backup_cancelled: me.magnum.melonds.dev
09-08 03:13:36.500 I am_proc_start: [0,30946,...,me.magnum.melonds.dev,top-activity,{me.magnum.melonds.dev/me.magnum.melonds.ui.romdetails.RomDetailsActivity}]
09-08 03:13:37.044 I wm_on_create_called: [...,me.magnum.melonds.ui.romdetails.RomDetailsActivity,performCreate]
```

</details>

## Multiple Recoveries While Sleeping (Overnight) Fix

On a subsequent test run, Android killed the sleeping emulator with `SIGKILL`; automatic checkpoint recovery succeeded. As recovery completed while the display remained off, the replacement session was recorded as awake, despite continuing to be asleep. Several hours later, scheduled backup killed that process again. The awake flag blocked automatic recovery, so the fallback prompt appeared.

Commit 6e5691b8 starts the normal sleep transition whenever emulation reaches the running state while the display is already off. The replacement session is paused and checkpointed, so another known non-user exit remains eligible for automatic restore.

Verification:

- Added four device-sleep transition policy tests; focused recovery tests and APK build passed.
- Installed practical master d3b4405d over `me.magnum.melonds.dev` without clearing app data.
- Installed package no longer advertises `ALLOW_BACKUP`; `bmgr backupnow` returns `Backup is not allowed`.
- On Thor, a running game was checkpointed with the display off, crashed, automatically restored and checkpointed while still off, crashed again, automatically restored again without a prompt, then resumed successfully on wake.

**Comment by corigne** [comment]
I've been testing this most recent build practically over the last 5 days with a single running emulation session, sleeping the device overnight while charging. The emulator session has been preserved seamlessly throughout the 5 days and I haven't experienced any other regressions.

While my PR branch represents this core work, my fork's `master` branch also includes in #1664 and #1589 for two reasons:
1. This PR relies on certain assumptions about the emulator's sleep/background behavior, and both PRs touch or approach sleep/background functionality. By merging them myself, I can prove they're logically compatible additions.
2. #1664 and #1589 are both valuable features, and merging them in allowed me to test them. I've been able to daily drive them for a while now with no major issues.

**Comment by corigne** [comment]
I think we're approaching a point where I've sussed out the obscure edges. I would like to see if you're happy with the prompt UX, and ask if there's a preference there or the recovery prompt is fine as-is. It's not pretty but it's also a fairly rare occurrence under normal circumstances.

**Comment by corigne** [comment]
I'm still daily driving this, and the last big issue I ran into, recently fixed, was recovered sessions sometimes not resuming the display renderer.

One last note, editing to avoid unnecessary notification spam...
I pre-emptively reached out to developer compliance for RetroAchievements explaining what I was doing and asking if it would be within the spirit of their developer compliance guidelines. The way I read them, I believe it's well and truly within the guidelines but I wanted to get a ye/nay before I completed the work. 

I've let the message sit with them for a bit, and they forwarded me to someone else on their team, which I'm still waiting for a response on. I'm still under the assumption that the behavior is acceptable and generally defended against abuse, but if you want to wait on their response I will post the information here, as I get it.
